delivery-service/ds/src/main.rs

```diff
@@ -0,0 +1,305 @@
+        bytes.extend_from_slice(&item.unwrap());
+    }
+    let info = match ClientInfo::decode(&mut Cursor::new(&bytes)) {
+        Ok(i) => i,
+        Err(_) => {
+            log::error!("Invalid payload for /clients/register\n{:?}", bytes);
+            return actix_web::HttpResponse::BadRequest().finish();
+        }
+    };
+    log::debug!("Registering client: {:?}", info);
+
+    let mut data = data.lock().unwrap();
+    let client_name = info.client_name.clone();
+    let old = data.clients.insert(info.id.clone(), info);
+    if old.is_some() {
+        return actix_web::HttpResponse::Conflict().finish();
+    }
+
+    actix_web::HttpResponse::Ok().body(format!("Welcome {}!\n", client_name))
+}
+
+/// Returns a list of clients with their names and IDs.
+#[get("/clients/list")]
+async fn list_clients(_req: HttpRequest, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    log::debug!("Listing clients");
+    let data = data.lock().unwrap();
+
+    // XXX: we could encode while iterating to be less wasteful.
+    let clients: Vec<ClientInfo> = data.deref().clients.values().map(|c| c.clone()).collect();
+    let mut out_bytes = Vec::new();
+    if clients.encode(&mut out_bytes).is_err() {
+        return actix_web::HttpResponse::InternalServerError().finish();
+    };
+    actix_web::HttpResponse::Ok().body(Body::from_slice(&out_bytes))
+}
+
+/// Resets the server state.
+#[get("/reset")]
+async fn reset(_req: HttpRequest, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    log::debug!("Resetting server");
+    let mut data = data.lock().unwrap();
+    let data = data.deref_mut();
+    data.clients.clear();
+    data.groups.clear();
+    actix_web::HttpResponse::Ok().finish()
+}
+
+/// Get the list of key packages for a given client `{id}`.
+/// This returns a serialised vector of `ClientKeyPackages` (see the `ds-lib` for
+/// details).
+#[get("/clients/key_packages/{id}")]
+async fn get_key_packages(
+    web::Path(id): web::Path<String>,
+    data: web::Data<Mutex<DsData>>,
+) -> impl Responder {
+    let data = data.lock().unwrap();
+
+    let id = match base64::decode_config(id, base64::URL_SAFE) {
+        Ok(v) => v,
+        Err(_) => return actix_web::HttpResponse::BadRequest().finish(),
+    };
+    log::debug!("Getting key packages for {:?}", id);
+
+    let client = data.clients.get(&id).unwrap();
+    actix_web::HttpResponse::Ok().body(Body::from_slice(
+        &client.key_packages.encode_detached().unwrap(),
+    ))
+}
+
+/// Send a welcome message to a client.
+/// This takes a serialised `Welcome` message and stores the message for all
+/// clients in the welcome message.
+#[post("/send/welcome")]
+async fn send_welcome(mut body: Payload, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    let mut bytes = web::BytesMut::new();
+    while let Some(item) = body.next().await {
+        bytes.extend_from_slice(&item.unwrap());
+    }
+    let welcome_msg = Welcome::decode(&mut Cursor::new(&bytes)).unwrap();
```

Comment:
Not sure how PoC-ish this should be in the end, but this would also be a good place to not panic :smile:

Reply:
yeah, I'll remove some of the unwraps and actually handle them